src/python/dxpy/sugar/context.py

```diff
@@ -0,0 +1,191 @@
+#   under the License.
+
+from __future__ import print_function
+import os
+import sys
+import tempfile
+import subprocess
+import contextlib
+import dxpy
+import json
+import logging
+
+LOG = logging.getLogger()
+
+class UserContext(object):
+    @dxpy.sugar.requires_worker_context
+    def __init__(self, api_token):
+        api_token = api_token.strip("\n")  # Python adds \n when reading from a file
+        self.user_secure_token = {"auth_token": api_token,
+                                 "auth_token_type": "Bearer"}
+        self.job_id = os.environ["DX_JOB_ID"]
+        self.job_security_context = json.loads(os.environ["DX_SECURITY_CONTEXT"])
+        self.job_workspace_id = os.environ["DX_WORKSPACE_ID"]
+
+    @dxpy.sugar.requires_worker_context
+    def __enter__(self):
+        proj = os.environ["DX_PROJECT_CONTEXT_ID"]
+        dxpy.set_job_id(None)
+        dxpy.set_security_context(self.user_secure_token)
+        dxpy.set_workspace_id(proj)
+        try:
+            dna_config_file = os.path.join(
+                os.environ["HOME"], ".dnanexus_config")
+            os.remove(dna_config_file)
+        except OSError:
+            LOG.info("As expected, .dnanexus_config not present.")
+        else:
+            LOG.error("Could not remove .dnanexus_config file.")
+        return self
+
+    @dxpy.sugar.requires_worker_context
+    def __exit__(self, type, value, traceback):
+        LOG.info("Restoring original Job context")
+        dxpy.set_job_id(self.job_id)
+        dxpy.set_security_context(self.job_security_context)
+        dxpy.set_workspace_id(self.job_workspace_id)
+
+@contextlib.contextmanager
+def set_env(**environ):
+    """
+    Context manager generator to temporarily set the subprocess environment variables.
+
+    Args:
+        environ (dict): Environment variable to set
+
+    Yields:
+        An environment with environment variables set as specified.
+        On exit, the environment will return to previous configuration.
+
+    Examples:
+        Usage 1: Set environment variable
+        # inside environment
+        >>> with set_env(PLUGINS_DIR=u'test/plugins'):
+        ...    "PLUGINS_DIR" in os.environ
+        True
+
+        # outside environment
+        >>> "PLUGINS_DIR" in os.environ
+        False
+
+        Usage 2: Unset environment variable
+        >>> with set_env(PYTHONPATH=''):
+        ...    print(os.environ["PYTHONPATH"])
+        <BLANKLINE>
+
+        Usage 3: Manipulate multiple variables
+        >>> myenv = {"PLUGINS_DIR": u'test/plugins', "PYTHONPATH": u'some/python/path'}
+        >>> with set_env(**myenv):
+        ...   print(os.environ["PLUGINS_DIR"])
+        ...   print(os.environ["PYTHONPATH"])
+        test/plugins
+        some/python/path
+    """
+    old_environ = dict(os.environ)
+    os.environ.update(environ)
+    try:
+        yield
+    finally:
+        os.environ.clear()
+        os.environ.update(old_environ)
+
+
+class cd:
+    """
```

Comment:
I'm curious why this one is done with a class vs @contextmanager decoration on function.

Reply:
It was done that way historically. I don't think rewriting this as a @contextmanager decoration would add much clarity as this class has explicit __enter__ and __exit__ functions.